src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
That's indeed the cleaner OO/encapsulated solution to the problem. Special return values are so C99 😉

How far do you go in making this 'idiot-proof'? If someone for instance does

```
NamedPipeTransaction a = pipe.transact();
// a has more data
NamedPipeTransaction b = pipe.transact();
b.readMoreData();
```

How do you deal with that case? Should the second pipe.transact fails because there's still an open transaction? Should it silently flush the pipe before writing the new request? ...

I think I'm trying to say that while this is a stylistic improvement it doesn't really prevent you from shooting yourself in the foot.